### PR TITLE
New function `count/0` for `Ecto.Query.API`

### PR DESCRIPTION
**Pull request by soteras**
### Objective
Create new function `count/0` for `Ecto.Query.API`

```
from(p in Post, select: count()) |> Repo.one

#Ecto.Query<from p in Post, select: count(:*)>                             

00:20:18.358 [debug] QUERY OK source="posts" db=0.4ms                      
SELECT count(*) FROM "posts" AS p0 []                                      
27                                   
iex(27)>   
```

### Information
There is a problem with `count/1` sometimes it can be slower.
See more #2595 

### Acceptance Criteria

- [x] New function in Ecto.Query.API
- [x] New functions Ecto.Query.Builder
- [x] Add new expression in postgres connection
- [x] Update docs
- [x] Add tests

### CODE REVIEW

**Comment by soteras** [comment]
I was a little confused in *Ecto.Query.Builder* about the return expressions. Could someone explain to me?

File: lib/ecto/query/api.ex [hunk 1]
```diff
@@ -158,6 +158,13 @@ defmodule Ecto.Query.API do
   """
   def count(value, :distinct), do: doc! [value, :distinct]
 
+  @doc """
+  Counts the entries in the table.
+
+      from p in Post, select: count()
+  """
+  def count, do: doc! ["*"]
```
Comment (josevalim) [inline lib/ecto/query/api.ex]: We should use `doc! []` since there are no arguments to `count`.

File: lib/ecto/query/builder.ex [hunk 2]
```diff
@@ -261,9 +261,15 @@ defmodule Ecto.Query.Builder do
     {expr, params_acc}
   end
 
+  def escape({:count, _, []}, type, params_acc, vars, env) do
+    expr = {:{}, [], [:count, [], [:*]]}
```
Comment (josevalim) [inline lib/ecto/query/builder.ex]: This should expand to `{:{}, [], [:count, [], []]}`.

Comment (josevalim) [reply lib/ecto/query/builder.ex]: The idea here is that `count(:*)` is a behaviour that is specific to SQL databases. We should only work with `count()` and the database can convert that to `count(:*)` if they find necessary.

**Comment by josevalim** [comment]
> I was a little confused in Ecto.Query.Builder about the return expressions. Could someone explain to me?

Which return expressions? From which function? :)

**Comment by josevalim** [comment]
This looks great! I have added some comments and we should be good to go!

**Comment by soteras** [comment]
Like this *{:{}, [], [:count, [], [:*]]}* i read some example but i didn't understand much. What are the meaning of each part of tuple? 

```
{
  :{} => ?
  [] => What can i put here?
  [:count, [], [:*]] => The second param i didn't understand
}
```

**Comment by josevalim** [comment]
Good question!

Since we are doing this at compile time, we need to return a data structured that, when compiled and executed will return Elixir AST. Therefore, what we need, is the **Escaped AST**.

The AST for `count()` is:

    iex> quote do: count()
    {:count, [], []}

The escaped AST is the one after we call `count()`:

    iex> Macro.escape quote do: count()
    {:{}, [], [:count, [], []]}

If we evaluate the escaped AST, we get the AST back:

    iex> Code.eval_quoted(Macro.escape quote do: count()) |> elem(0)
    {:count, [], []}

**Comment by soteras** [comment]
Now i understood. Thanks @josevalim

**Comment by soteras** [comment]
Done!!!

File: lib/ecto/adapters/postgres/connection.ex [hunk 3]
```diff
@@ -513,6 +513,8 @@ if Code.ensure_loaded?(Postgrex) do
       [?(, intersperse_map(elems, ?,, &expr(&1, sources, query)), ?)]
     end
 
+    defp expr({:count, _, []}, _sources, _query), do: ["count(*)"]
```
Comment (michalmuskala) [inline lib/ecto/adapters/postgres/connection.ex]: This doesn't need to be wrapped in a list.

**Comment by soteras** [comment]
@michalmuskala i did the changes and added this change in the mysql.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**